src/transaction/log_replication_atomic.hpp

```diff
@@ -98,6 +101,11 @@ namespace cublog
 
       log_lsa m_lowest_unapplied_lsa;
       mutable std::mutex m_lowest_unapplied_lsa_mutex;
+
+      /* Store the locked objects for DDL replication.
+       * Since multiple DDL operations can be executed within single transaction,
+       * more than one objects can be mapped to one transaction */
+      std::multimap <TRANID, OID> m_locked_objects;
```

Comment:
How about adding an assert `m_locked_classes.empty();` in the dtor, `~atomic_replicator()`?

Reply:
Agree! 
I should also consider cases where ATS performs DDL but dies before committing or aborting. Locks might remain in place. Perhaps ATS will create an abort log during recovery, but until then, it may continue to hold locks. Thanks.

Reply:
Oh, you're right. Then, `assert (m_locked_classes.empty())` can't be true always. Beyond that, transactions blocked by DDL by the replicator on PTS wait forever. Have we considered this? The transactions definitely are bound to be aborted.

Reply:
but, it's also true that we can't release DDL locks since it's not reverted. In this case, are all transactions just stuck until a new ATS arises?

Reply:
It seems like @hornetmj  had a similar question before. At that time, the answer was to wait for ATS to recover and then release the lock. Also, it was considered risky to release the lock while the DDL performed by ATS had not been completely applied to PTS. 
I can't think of any alternatives other than PTS transactions waiting..

Reply:
It seems not a simple issue, but I think we should address it. The infinite wait even can be cascaded to other transactions. If a Trx1 is blocked by a DDL transaction by ATS which will be never completed until an ATS is reconnected, other transactions blocked by the Trx1 are all blocked, and so on.

I assume we have an assumption that PTS can run even without ATS, but it seems to break the assumption easily. At least, we should have an idea about how to deal with this. For example, if a replicator is blocked too long (timeout) with a DDL, we can make disallow to access the table. In other words, if a transaction tries to access the table, just return an error. It, at least, allows transactions which don't access the table can be run.

There probably are better idea. We should define this problem and come up with a way to handle this case. It would be best prevent this problem.

Reply:
Since PTS transactions are read-only, they are unlikely to impact each other. It seems like we only need to be concerned about transactions blocked by the replicator. Given that this is not a simple issue, I've come up with a few ideas.

1. When a DDL start log arrives, it seems feasible to temporarily pause the replicator until replicator knows whether the DDL has been committed or aborted.
ex: 
  1) DDL start (LOG_SCHEMA_MODIFICATION_LOCK)
  2) Pause replicator
  3) DDL transaction complete (LOG_COMMIT, LOG_ABORT)
  4) Resume replicator.

In this case, a LOCK mechanism would likely be necessary, but it merely delays the replicator's application.

2. Alternatively, it might be possible to check both the connection status with ATS and lock mode, then report errors each time a read transaction in PTS access the object that is being locked by DDL.

I think we should create a separate issue to address this problem in more detail.

Reply:
> Since PTS transactions are read-only, they are unlikely to impact each other.

Oh, you're right. It's way less serious than I thought. We can just create an issue for this and move on for now.